File: Go/ExampleTests/common/catalogitem_string.go
package common

type CatalogItemString struct {
	Name string
	Price string
}

func NewCatalogItemStringFromSlice(v []string) CatalogItemString {
	s := CatalogItemString{}
	if len(v) > 0 { s.Name = v[0] }
	if len(v) > 1 { s.Price = v[1] }
	return s
}

// NewCatalogItemStringFromText builds from the text form, e.g. Money as "25 USD".
func NewCatalogItemStringFromText(text string) CatalogItemString {
	parts := RequireTokens(text, 2, "CatalogItem")
	return CatalogItemString{
		Name: parts[0],
		Price: parts[1],
	}
}

func (s CatalogItemString) String() string {
	return Token(s.Name) + " " + Token(s.Price)
}

func (s CatalogItemString) Equals(o CatalogItemString) bool {
	return DNCEqual(s.Name, o.Name) &&
		DNCEqual(s.Price, o.Price)
}

func EqualCatalogItemStringSlices(a, b []CatalogItemString) bool {
	if len(a) != len(b) { return false }
	for i := range a {
		if !a[i].Equals(b[i]) { return false }
	}
	return true
}

File: Rust/ExampleTests/common/catalogitem_string.rs
#![allow(dead_code, unused_imports, unused_variables)]

#[derive(Debug, Clone, Default)]
pub struct CatalogItemString {
    pub name: String,
    pub price: String,
}

impl CatalogItemString {
    pub fn from_vec(v: &[&str]) -> Self {
        Self {
            name: v.get(0).copied().unwrap_or("").to_string(),
            price: v.get(1).copied().unwrap_or("").to_string(),
        }
    }

    /// Builds from the text form, e.g. Money as "25 USD".
    pub fn from_text(text: &str) -> Self {
        let parts = crate::common::tokens::require(text, 2, "CatalogItem");
        Self {
            name: parts[0].clone(),
            price: parts[1].clone(),
        }
    }
}

impl std::fmt::Display for CatalogItemString {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", [crate::common::tokens::token(&self.name), crate::common::tokens::token(&self.price)].join(" "))
    }
}

impl PartialEq for CatalogItemString {
    fn eq(&self, other: &Self) -> bool {
        crate::common::dnc_equal(&self.name, &other.name)
            && crate::common::dnc_equal(&self.price, &other.price)
    }
}
